cpp/ql/src/Security/CWE/CWE-311/CleartextTransmission.ql

```diff
@@ -0,0 +1,104 @@
+/**
+ * @name Cleartext transmission of sensitive information
+ * @description Transmitting sensitive information across a network in
+ *              cleartext can expose it to an attacker.
+ * @kind path-problem
+ * @problem.severity warning
+ * @security-severity 7.5
+ * @precision medium
+ * @id cpp/cleartext-transmission
+ * @tags security
+ *       external/cwe/cwe-319
+ */
+
+import cpp
+import semmle.code.cpp.security.SensitiveExprs
+import semmle.code.cpp.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/**
+ * A function call that sends or receives data over a network.
+ */
+abstract class NetworkSendRecv extends FunctionCall {
+  /**
+   * Gets the expression for the socket or similar object used for sending or
+   * receiving data.
+   */
+  abstract Expr getSocketExpr();
+
+  /**
+   * Gets the expression for the buffer to be sent from / received into.
+   */
+  abstract Expr getDataExpr();
+}
+
+/**
+ * A function call that sends data over a network.
+ *
+ * note: functions such as `write` may be writing to a network source or a file. We could attempt to determine which, and sort results into `cpp/cleartext-transmission` and perhaps `cpp/cleartext-storage-file`. In practice it usually isn't very important which query reports a result as long as its reported exactly once.
+ */
+class NetworkSend extends NetworkSendRecv {
```

Comment:
Could we not replace this whole class with `RemoteFlowSinkFunction`?

Reply:
Yes!  Ish...

First, this class is a _call to_ a `RemoteFlowSinkFunction`/`RemoteFlowSourceFunction`, not a `RemoteFlowSinkFunction`/`RemoteFlowSourceFunction` itself, so I'll still have to wrap those - unless there's a cleaner way to wire it up.  But this does avoid duplicating the list of functions, which is a good thing.

Second, `RemoteFlowSourceFunction` doesn't have an equivalent of `getSocketExpr()` for ruling out reads from standard input.  We could have `RemoteFlowSourceFunction` expose this information or even do the work to rule out these sources itself?

Third, `RemoteFlowSourceFunction` also includes `getEnv` (I'm not sure why, I thought the whole point was to be able to specify sources a bit more precisely).  So ideally we'd use something finer that doesn't include that (but `Recv` is currently private).

Reply:
> Third, `RemoteFlowSourceFunction` also includes `getEnv` (I'm not sure why, I thought the whole point was to be able to specify sources a bit more precisely). So ideally we'd use something finer that doesn't include that (but `Recv` is currently private).

I think `getEnv` is only included in `LocalFlowSourceFunction`, right? That's at least what I see [here](https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/models/implementations/Getenv.qll#L11).

Reply:
> First, this class is a _call to_ a `RemoteFlowSinkFunction`/`RemoteFlowSourceFunction`, not a `RemoteFlowSinkFunction`/`RemoteFlowSourceFunction` itself, so I'll still have to wrap those - unless there's a cleaner way to wire it up. But this does avoid duplicating the list of functions, which is a good thing.

We have `RemoteFlowSource` wrapping `RemoteFlowSourceFunction` in `semmle.code.cpp.models.interfaces.FlowSource`, but no equivalent for sinks.

Reply:
> Could we replace this whole class with `RemoteFlowSinkFunction`?
> Could we replace this whole file with RemoteFlowSourceFunction?

I've updated the code to (more-or-less) do this.

The following are considered a `RemoteFlowSourceFunction` (`#` = not already modelled by `NetworkRecv`): `recv`, `recvfrom`, `recvmsg`, `read`, `pread`, `readv`, `preadv`, `preadv2#`, `gets#`, `fgets#`, `fgetws#`, `getdelim#`, `getwdelim#`, `__getdelim#`, `fread#`.  Some of the new functions read from a socket, some from a `FILE *`, and I'm a bit concerned that we'll get overlaps with `cpp/cleartext-storage-file` from the latter.  They're differently structured queries though so I'm not sure this will really be an issue in practice.  Probably best to fix it when we see it.

The following are considered a `RemoteFlowSinkFunction` (all already modelled by `NetworkSend`): `send`, `sendto`, `sendmsg`, `write`, `writev`, `pwritev`, `pwritev2`.  So no problems there.

> We have `RemoteFlowSource` wrapping `RemoteFlowSourceFunction` in `semmle.code.cpp.models.interfaces.FlowSource`, but no equivalent for sinks.

This looks promising, but it's in the `security.FlowSources` lib which conflicts with `dataflow.TaintTracking` (and as you point out `RemoteFlowSink` would need to be added).  This PR has enough in it that I won't do this change here, but if you feel strongly I can do it as a follow-up?

> > Second, `RemoteFlowSourceFunction` doesn't have an equivalent of `getSocketExpr()` for ruling out reads from standard input. We could have `RemoteFlowSourceFunction` expose this information or even do the work to rule out these sources itself?

> This seems like a thing we would want to do in `RemoteFlowSourceFunction`, yeah.

It turns out `RemoteFlowSourceFunction` can't rule out stdin/stdout by itself, as it's modelling the function not the call and thus has no idea of parameter values.  We could potentially do this work in an improved `RemoteFlowSource` class at some point.  But for now I've added a `hasSocketInput` predicate to the two models so that users can tell which parameter is the socket (if any).

> I think `getEnv` is only included in `LocalFlowSourceFunction`, right? 

Correct, thanks.

---

TL;DR: I've updated the query to use the `RemoteFlowSourceFunction` and `RemoteFlowSinkFunction` models.  I believe all points above have been addressed, apart from the possible follow-up PR improving `RemoteFlowSource`.

Reply:
One of the reasons we generally prefer modelling `Function` rather than `Call` is that a `Function` model also applies directly to IR dataflow, whereas a `Call` model has to be translated. Of course, a `Call` model can have more details. The hybrid you've arrived at here LGTM.

Reply:
It might be better (at least, less code) to join the `NetworkSend` and `NetworkReceive` parts at the `Function` level, then look at calls further down.  I'll have a go at a refactor...

Reply:
It doesn't make much difference to code quantity or clarity IMO.  The awkward bit remains converting from a `FunctionInput` / `FunctionOutput` to an `Expr` or `DataFlow::Node`.  Is there a reason we don't have predicates to do this in `FunctionInputsAndOutputs.qll`?

Reply:
AFAIK there is no reason why we don't have this logic centralized in `FunctionInputsAndOutputs.qll`, no. I'll open an issue for it.